includes/Story_Renderer/HTML.php

```diff
@@ -111,82 +69,41 @@ public function __construct( Story $story ) {
 	 * @return string The complete HTML markup for the story.
 	 */
 	public function render() {
-		$markup = self::DEFAULT_DOCTYPE . $this->story->get_markup();
-		$markup = $this->adapt_encoding( $markup );
+		$markup = $this->story->get_markup();
 		$markup = $this->replace_html_head( $markup );
-		$markup = $this->remove_noscript_amp_boilerplate( $markup );
 
-		$this->document = $this->load_html( $markup );
+		// TODO: What if transformation failed?
+		$document = Document::fromHtml( $markup, get_bloginfo( 'charset' ) );
+
+		// This  should never actually happen.
+		if ( ! $document ) {
+			wp_die(
+				esc_html__( 'There was an error generating the web story, probably because of a server misconfiguration. Try contacting your hosting provider or open a new support request.', 'web-stories' ),
+				esc_html__( 'Web Stories', 'web-stories' ),
+				[
+					'response'  => 500,
+					'link_url'  => esc_url( __( 'https://wp.stories.google/', 'web-stories' ) ),
+					'link_text' => esc_html__( 'Contact Support', 'web-stories' ),
+				]
+			);
+		}
 
-		// Run all further transformations on the DOMDocument.
+		$this->document = $document;
+
+		// Run all further transformations on the Document instance.
 
-		$this->add_noscript_amp_boilerplate();
 		$this->transform_html_start_tag();
 		$this->transform_a_tags();
 		$this->insert_analytics_configuration();
 
 		$this->add_poster_images();
-		$this->add_publisher_logo();
-
-		return trim( (string) $this->document->saveHTML() );
-	}
 
-	/**
-	 * Adapt the encoding of the content.
-	 *
-	 * @link https://github.com/ampproject/amp-wp/blob/a393acf701e8e44d80225affed99d528ee751cb9/lib/common/src/Dom/Document.php
-	 *
-	 * @since 1.0.0
-	 *
-	 * @param string $markup Source content to adapt the encoding of.
-	 *
-	 * @return string Adapted content.
-	 */
-	protected function adapt_encoding( $markup ) {
-		$encoding = self::UNKNOWN_ENCODING;
-
-		if ( function_exists( 'mb_detect_encoding' ) ) {
-			$encoding = mb_detect_encoding( $markup, self::ENCODING_DETECTION_ORDER, true );
+		if ( ! defined( '\AMP__VERSION' ) ) {
```

Comment:
This way, if the AMP plugin is installed, all sanitization and optimization is delegated to the AMP plugin.

TODO: Only do that if `AMP__VERSION` is higher than or equal to what's included in the web Stories plugin.

Reply:
@westonruter does the above sound reasonable to you?

Reply:
Yes, I think that makes sense.